### PR TITLE
type-and-swap: add openapi.yaml

### PR DESCRIPTION
**Pull request by vuquang23**
Add api design for feature type-and-swap

### CODE REVIEW

File: openapi/type-and-swap/openapi.yaml [hunk 1]
```diff
@@ -0,0 +1,637 @@
...
+        '400':
+          description: |-
+            Bad Request
+            - Missing required fields.
+            - ChainId is invalid.
+          content:
+            application/json:
+              schema:
+                type: object
+                x-examples:
```
Comment (anhvietnguyennva) [inline openapi/type-and-swap/openapi.yaml:185]: Is it redundant?

File: openapi/type-and-swap/openapi.yaml [hunk 2]
```diff
@@ -0,0 +1,637 @@
...
+  /api/v1/favorite-pairs:
+    post:
+      summary: Add new favorite trading pair
+      operationId: post-api-v1-favorite
+      requestBody:
+        content:
+          application/json:
+            schema:
+              type: object
+              x-examples:
```
Comment (anhvietnguyennva) [inline openapi/type-and-swap/openapi.yaml:279]: Is it redundant?

File: openapi/type-and-swap/openapi.yaml [hunk 3]
```diff
@@ -0,0 +1,637 @@
...
+      description: ''
+      x-internal: false
+      responses:
+        '200':
+          description: OK
+          content:
+            application/json:
+              schema:
+                type: object
+                x-examples:
```
Comment (anhvietnguyennva) [inline openapi/type-and-swap/openapi.yaml:316]: Is it redundant?

File: openapi/type-and-swap/openapi.yaml [hunk 4]
```diff
@@ -0,0 +1,637 @@
...
+        '400':
+          description: |-
+            Bad Request
+            - Missing required fields.
+            - ChainId is invalid.
+          content:
+            application/json:
+              schema:
+                type: object
+                x-examples:
```
Comment (anhvietnguyennva) [inline openapi/type-and-swap/openapi.yaml:359]: Is it redundant?

File: openapi/type-and-swap/openapi.yaml [hunk 5]
```diff
@@ -0,0 +1,637 @@
...
+                    details: []
+        '404':
+          description: |-
+            Not Found
+            - tokenIn, tokenOut not found
+          content:
+            application/json:
+              schema:
+                type: object
+                x-examples:
```
Comment (anhvietnguyennva) [inline openapi/type-and-swap/openapi.yaml:402]: Is it redundant?

File: openapi/type-and-swap/openapi.yaml [hunk 6]
```diff
@@ -0,0 +1,637 @@
...
+                    errorEntities:
+                      - tokenIn
+                    details: []
+        '500':
+          description: Internal Server Error
+          content:
+            application/json:
+              schema:
+                type: object
+                x-examples:
```
Comment (anhvietnguyennva) [inline openapi/type-and-swap/openapi.yaml:440]: Is it redundant?

File: openapi/type-and-swap/openapi.yaml [hunk 7]
```diff
@@ -0,0 +1,637 @@
...
+                    details: []
+    delete:
+      summary: Remove favorited trading pair
+      operationId: delete-api-v1-favorite
+      requestBody:
+        content:
+          application/json:
+            schema:
+              type: object
+              x-examples:
```
Comment (anhvietnguyennva) [inline openapi/type-and-swap/openapi.yaml:481]: Is it redundant?

File: openapi/type-and-swap/openapi.yaml [hunk 8]
```diff
@@ -0,0 +1,637 @@
...
+        description: ''
+      description: ''
+      responses:
+        '200':
+          description: OK
+          content:
+            application/json:
+              schema:
+                type: object
+                x-examples:
```
Comment (anhvietnguyennva) [inline openapi/type-and-swap/openapi.yaml:517]: Is it redundant?

File: openapi/type-and-swap/openapi.yaml [hunk 9]
```diff
@@ -0,0 +1,637 @@
...
+        '400':
+          description: |-
+            Bad Request
+            - Missing required fields.
+            - ChainId is invalid.
+          content:
+            application/json:
+              schema:
+                type: object
+                x-examples:
```
Comment (anhvietnguyennva) [inline openapi/type-and-swap/openapi.yaml:558]: Is it redundant?

File: openapi/type-and-swap/openapi.yaml [hunk 10]
```diff
@@ -0,0 +1,637 @@
...
+                    errorEntities:
+                      - wallet
+                    details: []
+        '500':
+          description: Internal Server Error
+          content:
+            application/json:
+              schema:
+                type: object
+                x-examples:
```
Comment (anhvietnguyennva) [inline openapi/type-and-swap/openapi.yaml:599]: Is it redundant?

File: openapi/type-and-swap/openapi.yaml [hunk 11]
```diff
@@ -0,0 +1,637 @@
...
+                  - details
+              examples:
+                example-1:
+                  value:
+                    code: 5000
+                    message: Internal Server Error
+                    errorEntities: []
+                    details: []
+    parameters: []
+components:
```
Comment (anhvietnguyennva) [inline openapi/type-and-swap/openapi.yaml:634]: pls remove unused code

File: openapi/type-and-swap/openapi.yaml [hunk 12]
```diff
@@ -0,0 +1,637 @@
...
+      summary: Get suggested pairs by user text input
+      tags: []
+      responses:
+        '200':
+          description: OK
+          content:
+            application/json:
+              schema:
+                type: object
+                x-examples:
```
Comment (anhvietnguyennva) [inline openapi/type-and-swap/openapi.yaml:27]: Is it redundant?

File: openapi/type-and-swap/openapi.yaml [hunk 13]
```diff
@@ -0,0 +1,637 @@
...
+            type: string
+          in: query
+          name: chainId
+          description: Blockchain ChainID
+          required: true
+      description: ''
+    parameters: []
+  /api/v1/favorite-pairs:
+    post:
+      summary: Add new favorite trading pair
```
Comment (anhvietnguyennva) [inline openapi/type-and-swap/openapi.yaml:272]: pls add tags here

File: openapi/type-and-swap/openapi.yaml [hunk 14]
```diff
@@ -0,0 +1,637 @@
...
+                  - details
+              examples:
+                example-1:
+                  value:
+                    code: 5000
+                    message: Internal Server Error
+                    errorEntities: []
+                    details: []
+    delete:
+      summary: Remove favorited trading pair
```
Comment (anhvietnguyennva) [inline openapi/type-and-swap/openapi.yaml:474]: pls add tags here

**Review by anhvietnguyennva — COMMENTED** [review]
@vuquang23 Please help me change the folder name to `tns`

File: openapi/type-and-swap/openapi.yaml (hunk 1 above)
Comment (vuquang23) [reply openapi/type-and-swap/openapi.yaml:185]: it defines schema of response object. I am not sure whether it's redundant?

**Review by anhvietnguyennva — APPROVED** [review]

File: openapi/tns/openapi.yaml [hunk 15]
```diff
@@ -0,0 +1,518 @@
...
+    post:
+      summary: Add new favorite trading pair
+      operationId: post-api-favorite
+      requestBody:
+        content:
+          application/json:
+            schema:
+              type: object
+              properties:
+                wallet:
```
Comment (hanguyenk) [inline openapi/tns/openapi.yaml:239]: I think we can have format validation for such fields:
```golang
func ValidateAddress(address string) bool {
	re := regexp.MustCompile("^0x[0-9a-fA-F]{40}$")

	return re.MatchString(address)
}
```

**Review by hanguyenk — APPROVED** [review]
LGTM

**Pull request merged**